### PR TITLE
Add pre-release testing for OSA

### PR DESCRIPTION
**Pull request by mjudeikis**
Add a variable to replace multiple versions in the scripts. 
Add rhel pre-release secret passing to ansible code. This will need to be only for pre-release builds of rhel. 

cc: @openshift/sig-azure

### CODE REVIEW

File: sjb/config/test_cases/azure_build_node_image_rhel_310.yml [hunk 1]
```diff
@@ -10,17 +10,23 @@ sync:
 parameters:
 - name: "AZURE_VM_SIZE"
   default_value: "Standard_D2s_v3"
+- name: "OPENSHIFT_RELEASE"
+  default_value: "3.10"
```
Comment (kargakis) [inline sjb/config/test_cases/azure_build_node_image_rhel_310.yml]: The job is already versioned so we don't need these variables.

File: sjb/config/test_cases/azure_build_node_image_rhel_310.yml [hunk 2]
```diff
@@ -10,17 +10,23 @@ sync:
...
+- name: "OPENSHIFT_RELEASE_LATEST"
```
Comment (kargakis) [inline sjb/config/test_cases/azure_build_node_image_rhel_310.yml]: I am not sure we need this parameter. Playbooks are already versioned based on the branch (we know that playbooks run from release branches are not latest).

Comment (mjudeikis) [reply sjb/config/test_cases/azure_build_node_image_rhel_310.yml]: as discusses on irc will move to maintain this in the ansible code via release branches.

**Comment by mjudeikis** [comment]
@kargakis  I know we discussed this on chat, but its big pita for rhel builds to run without knowing which version they are running. Idea is that our ansible playbooks is version agnostic, so they are capable to build centos and rhel for all releases. 

If we stick with "branch code versioning" we will soon end-up maintaining  versions of code (even if its just 3 places in 2 repos where we need to change few lines of version per branches), it will be poor man solution as we will need to make 2 repos 3 releases = 6 PR if we want to change anything... Not the place I want to be in.

As much as I dont like this I would better improve our openshift-ansible to build based on the variable we pass from here, and do maintenance in one place and not over 2-3 separate repositories :/ 

Long story short problem is our RPM channels and how we treat some of the logic in openshift-ansible. To be safe we need these in our node build scripts
```
    openshift_image_tag: "v3.10"
    openshift_version: "3.10"
    openshift_release: "3.10"
```

File: sjb/config/test_cases/azure_build_node_image_centos_311.yml [hunk 3]
```diff
@@ -0,0 +1,112 @@
+---
```
Comment (kargakis) [inline sjb/config/test_cases/azure_build_node_image_centos_311.yml]: This shouldn't be `*centos_311.yml` but `*centos.yml` like all other master jobs.

File: sjb/config/test_cases/azure_build_node_image_centos_311.yml [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+provision:
+  os: "rhel"
+  stage: "base"
+  provider: "aws"
+sync:
+  - openshift,openshift-ansible=release-3.11
```
Comment (kargakis) [inline sjb/config/test_cases/azure_build_node_image_centos_311.yml]: There is no such branch (yet). Use master.

File: sjb/config/test_cases/azure_build_node_image_centos_311.yml [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        -e IMAGE_NAME=$IMAGE_NAME \
+        -e CLUSTER_DEFINITION=examples/openshift.json \
+        -e TIMEOUT=30m \
+        -e CLEANUP_ON_EXIT=true \
+        -e REGIONS=eastus \
+        -e CLIENT_ID=$AZURE_CLIENT_ID \
+        -e CLIENT_SECRET=$AZURE_SECRET \
+        -e TENANT_ID=$AZURE_TENANT \
+        -e SUBSCRIPTION_ID=$AZURE_SUBSCRIPTION_ID \
+        registry.svc.ci.openshift.org/azure/acs-engine-tests:v$OPENSHIFT_RELEASE make test-openshift
```
Comment (kargakis) [inline sjb/config/test_cases/azure_build_node_image_centos_311.yml]: Use 3.10 here

File: sjb/config/test_cases/azure_build_node_image_centos_311.yml (hunk 4 above)
Comment (mjudeikis) [reply sjb/config/test_cases/azure_build_node_image_centos_311.yml]: tempted to delete all these for now. I was not expting to get config for jenkis so fast :D

File: sjb/config/test_cases/azure_build_node_image_centos_311.yml [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    parameters:
+      - JOB_SPEC
+      - AZURE_VM_SIZE
+      - OPENSHIFT_RELEASE
+
+  - type: "host_script"
+    title: "determine the resource group"
+    script: |-
+      echo "RESOURCE_GROUP=ci-$JOB_NAME-$BUILD_NUMBER" | ssh -F ./.config/origin-ci-tool/inventory/.ssh_config openshiftdevel "cat >>/etc/environment"
+      echo "IMAGE_NAME=centos7-$OPENSHIFT_RELEASE-$(TZ=Etc/UTC date +%Y%m%d%H%M)" | ssh -F ./.config/origin-ci-tool/inventory/.ssh_config openshiftdevel "cat >>/etc/environment"
```
Comment (kargakis) [inline sjb/config/test_cases/azure_build_node_image_centos_311.yml]: Use `master` instead of OPENSHIFT_RELEASE

File: sjb/config/test_cases/azure_build_node_image_rhel_311.yml [hunk 7]
```diff
@@ -0,0 +1,121 @@
+---
+provision:
+  os: "rhel"
+  stage: "base"
+  provider: "aws"
+sync:
+  - openshift,openshift-ansible=release-3.11
```
Comment (kargakis) [inline sjb/config/test_cases/azure_build_node_image_rhel_311.yml]: master

File: sjb/config/test_cases/azure_build_node_image_rhel_311.yml [hunk 8]
```diff
@@ -0,0 +1,121 @@
+---
```
Comment (kargakis) [inline sjb/config/test_cases/azure_build_node_image_rhel_311.yml]: Need to drop the 311 suffix

File: sjb/config/test_cases/azure_build_node_image_rhel_311.yml [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    parameters:
+      - JOB_SPEC
+      - AZURE_VM_SIZE
+      - OPENSHIFT_RELEASE
+
+  - type: "host_script"
+    title: "determine the resource group"
+    script: |-
+      echo "RESOURCE_GROUP=ci-$JOB_NAME-$BUILD_NUMBER" | ssh -F ./.config/origin-ci-tool/inventory/.ssh_config openshiftdevel "cat >>/etc/environment"
+      echo "IMAGE_NAME=rhel7-$OPENSHIFT_RELEASE-$(TZ=Etc/UTC date +%Y%m%d%H%M)" | ssh -F ./.config/origin-ci-tool/inventory/.ssh_config openshiftdevel "cat >>/etc/environment"
```
Comment (kargakis) [inline sjb/config/test_cases/azure_build_node_image_rhel_311.yml]: master

File: sjb/config/test_cases/azure_build_node_image_rhel_311.yml [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        -e IMAGE_NAME=$IMAGE_NAME \
+        -e CLUSTER_DEFINITION=examples/openshift.json \
+        -e TIMEOUT=30m \
+        -e CLEANUP_ON_EXIT=true \
+        -e REGIONS=eastus \
+        -e CLIENT_ID=$AZURE_CLIENT_ID \
+        -e CLIENT_SECRET=$AZURE_SECRET \
+        -e TENANT_ID=$AZURE_TENANT \
+        -e SUBSCRIPTION_ID=$AZURE_SUBSCRIPTION_ID \
+        registry.svc.ci.openshift.org/azure/acs-engine-tests:v$OPENSHIFT_RELEASE make test-openshift
```
Comment (kargakis) [inline sjb/config/test_cases/azure_build_node_image_rhel_311.yml]: 3.10

**Comment by mjudeikis** [comment]
/hold cancel
ptal @kwoodson @kargakis @jim-minter   This passes token to be backed into the image for RHEL build which needs to be built from aws registry. This particular code needs to be removed when going GA and moving jobs.

**Comment by kwoodson** [comment]
@mjudeikis, I noticed you have the ...311.xml files but not the corresponding  ....yml files.  Are they missing or did I miss something?

File: sjb/config/test_cases/azure_build_node_image_centos_39.yml [hunk 11]
```diff
@@ -51,6 +54,7 @@ actions:
         -e "openshift_azure_storage_account_ns=images" \
         -e "openshift_azure_container=images" \
         -e "openshift_azure_vm_size=$AZURE_VM_SIZE" \
+        -e "openshift_azure_release=$OPENSHIFT_RELEASE" \
```
Comment (kargakis) [inline sjb/config/test_cases/azure_build_node_image_centos_39.yml]: Is this needed?

Comment (mjudeikis) [reply sjb/config/test_cases/azure_build_node_image_centos_39.yml]: ack.

File: sjb/config/test_cases/azure_build_node_image_rhel_310.yml [hunk 12]
```diff
@@ -31,6 +34,10 @@ actions:
   - type: "host_script"
     title: "move secrets to remote host"
     script: |-
+      set +x
+      export KUBECONFIG=/var/lib/jenkins/.azure/kubeconfig
+      echo $(oc get secret aws-reg-token --template={{.data.token}} |base64 -d) > /var/lib/jenkins/.azure/registry
```
Comment (kargakis) [inline sjb/config/test_cases/azure_build_node_image_rhel_310.yml]: I assume we will use this eventually for running PR tests. If `oc get secret` fails then the token will get overwritten and jobs running in parallel beyond this step will not have a token to use?

Comment (mjudeikis) [reply sjb/config/test_cases/azure_build_node_image_rhel_310.yml]: Whats true. How unreliable is our CI cluster for this? Do we need to build some contingency in this now? I would like to assume perfect world for now and if we see a lot of flakes we can add some checks.

Comment (kargakis) [reply sjb/config/test_cases/azure_build_node_image_rhel_310.yml]: The CI cluster is mostly reliable but the minimal fix for avoiding this flake comes at no cost.
```bash
NEW_TOKEN=$(oc get secret aws-reg-token --template={{.data.token}} | base64 -d)
if [[ -n $NEW_TOKEN ]]; then
  echo $NEW_TOKEN > /var/lib/jenkins/.azure/registry
fi
```

File: sjb/config/test_cases/azure_build_node_image_rhel_310.yml [hunk 13]
```diff
@@ -31,6 +34,13 @@ actions:
   - type: "host_script"
     title: "move secrets to remote host"
     script: |-
+      set +x
+      export KUBECONFIG=/var/lib/jenkins/.azure/kubeconfig
+      NEW_TOKEN=$(oc get secret aws-reg-token --template={{.data.token}} | base64 -d)
+      if [[ -n $NEW_TOKEN ]]; then
+        echo $NEW_TOKEN > /var/lib/jenkins/.azure/registry
```
Comment (jim-minter) [inline sjb/config/test_cases/azure_build_node_image_rhel_310.yml]: don't write to the jenkins filesystem, it introduces a (tiny) race - echo the token up to the box via ssh echo.  Think about the filename you choose - make it descriptive

File: sjb/config/test_cases/azure_build_node_image_rhel_310.yml [hunk 14]
```diff
@@ -31,6 +34,13 @@ actions:
   - type: "host_script"
     title: "move secrets to remote host"
     script: |-
+      set +x
```
Comment (mjudeikis) [inline sjb/config/test_cases/azure_build_node_image_rhel_310.yml]: Move to file-less way to avoid race conditions:
 ssh -F ./.config/origin-ci-tool/inventory/.ssh_config openshiftdevel "echo 'location_url=${location_url:-}' >> /etc/environment"
or 
>/data/src/github.com/openshift/release/cluster/test-deploy/azure/registry

File: sjb/config/test_cases/azure_build_node_image_rhel_310.yml [hunk 15]
```diff
@@ -51,7 +65,9 @@ actions:
         -e "openshift_azure_storage_account_ns=images" \
         -e "openshift_azure_container=images" \
         -e "openshift_azure_vm_size=$AZURE_VM_SIZE" \
-        playbooks/azure/openshift-cluster/build_node_image.yml
+        -e "openshift_azure_dev_registry_token=$(cat /data/src/github.com/openshift/release/cluster/test-deploy/azure/registry)" \
```
Comment (jim-minter) [inline sjb/config/test_cases/azure_build_node_image_rhel_310.yml]: add a comment that this line should be removed at GA

**Comment by jim-minter** [comment]
2 nits then will merge

**Comment by jim-minter** [comment]
@mjudeikis merged - please will you update the jenkins jobs?

**Comment by mjudeikis** [comment]
@jim-minter done and running
